### PR TITLE
[Ready for review] Sketcher: Circle to circle distance constraints

### PR DESCRIPTION
**Pull request by FlachyJoe**
I'm trying to continue the work of @davidosterberg started here: #5439

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [796702985](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/796702985) was triggered at [2f73584](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/2f735847cffa3956c29c709c773a670138a37bcc). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [797199705](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/797199705) was triggered at [fd9e289](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/fd9e28941f2d3df158277c2ab6ba9d550b436e88). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by FlachyJoe** [comment]
@abdullahtahiriyo could you please tell me how to deal with constraint's label placement ?

**Comment by abdullahtahiriyo** [comment]
https://github.com/FreeCAD/FreeCAD/blob/1a0801a23f917f534d4dabe8c956df4cd206e2a2/src/Mod/Sketcher/Gui/EditModeConstraintCoinManager.cpp#L156

There you have all the positioning of constraints. Take a look at one having a label, such as radius.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [797380268](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/797380268) was triggered at [9ba941f](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/9ba941fe7b1db29235e46081e5ea2b03a3adf4f8). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by abdullahtahiriyo** [comment]
I took a look to the code.

IMO overloading the "Distance" Sketcher constraint for a concentric constraint is a hack. 

At solver level this implementation just sets a difference between radii. There is no guarantee of concentric.

Conceptually, any Python power user would expect that creating a distance between two geoids belonging to two circles would enforce such a distance also when the circles are not concentric.

If this Sketcher constrain would be reimplemented as "Concentric", I would argue that the coincident constraint is what makes the two circles concentric. I have problems to see the added value of "Concentric".

If the intention is to provide a distance between circles (which would lead to an offset between edges when their centers are coincident), then the solver part needs to be reworked to consider the centers of the circles in addition to the radii. So, basically the problem to be solved by the solver is that of a distance between edges of circles with centers at any positions. This problem will lead to an offset when the centers are concentric.

**Comment by abdullahtahiriyo** [comment]
> But how to define a distance between two circle? Which points to take into consideration? In my opinion there is a unique case where two circle has a specific distance between them and it's the concentric one.

umm....not sure how you are looking at it... let's try to agree how to look at it.

Above I consider the distance between two circles to be the minimum distance between the two edges of a circle. So, drawing a line from center1 to center 2 and deducting the two radii is the distance between circles (this is the distance along the common perpendicular to the edges of both circles).

This is also the case to point to line distance (the distance along the perpendicular to the line).

Coincidentally, as I understand it, you are also referring to the same. When the two circles are concentric, there are infinite paths to go from the edge of circle 1 to the edge of circle 2, but you are considering the distance along the perpendicular, which happens to be the difference between radii.

Do you see it differently?

**Comment by FlachyJoe** [comment]
Let's define circle to circle distance as the minimal one.
Now we have two possibility:
1. write the solver code for this and use the Distance constraint.
2. Create a Radius Difference (or whatever we can name it) constraint based on the current code without the concentric requirement.

**Comment by abdullahtahiriyo** [comment]
I think it makes sense to overload the sketcher distance constraint to a case of circle to circle (1).

The problem with (2) is twofold: (a) I do not think it makes much practical sense a constraint "radius difference", (b) that constraint would not have support at sketcher level, so we would need to create a new sketcher constraint to map it. 

Now, you have invested quite so work in this. Do you feel like writing the solver constraint of (1)? 

You need a new constraint taking as parameters both center points, both radii and a fixed value (the offset or distance). You may use derivector to define a Vector between the center points. Your error function is the length of derivector minus both radii minus the offset. You can extract the gradients directly from derivector.

This is an example of how to use derivector:
https://github.com/FreeCAD/FreeCAD/blob/1a0801a23f917f534d4dabe8c956df4cd206e2a2/src/Mod/Sketcher/App/planegcs/Constraints.cpp#L2519

Give it a look if you need help mention me.

**Comment by FlachyJoe** [comment]
@abdullahtahiriyo there is no sense to add circle to circle distance without line to circle and point to circle.
If I understand well I have 3 constraints to define in planegcs/Constraints.cpp and I can use ConstraintP2LDistance as template.

**Comment by abdullahtahiriyo** [comment]
I am ok if you want to add circle-to-circle or line to circle or both. 

My advice is: if you can make several of them of a single solver constraint go for it. If you need more than one, do a PR per solver constraint. I have no preference for which one calls firsties. That is going to be better for you too. If now you have all set for circle-to-circle, I would consider finishing it and then undertaking the other. But, of course, you code, you get to code what you want to :)

There are two ways to code solver constraints. One is the Constraint P2LDistance way (without derivector). The other one is the one I linked above (with derivector). With derivector you do not have to care about the gradient calculation (you write one function and derivector takes care of providing the differentials for the gradient. Without derivector you need to provide both the error and the gradient functions yourself.

Derivector may seem a little bit odd at first, but it is very easy to build geometric vectors. My advice is to go the derivector way. But if you feel very much comfortable with calculating functions and derivatives, you may want to go the other way. The other way has more checking to be done. But I will review whatever path you decide to take.

Point to circle and circle to circle are likely to be doable with a same constraint, where the radius of the point is zero. But this you will see as you code it (perhaps I am not anticipating an issue).

**Comment by PaddleStroke** [comment]
Great to see this has been picked up.
I agree that a circle to circle 'general' distance is better than concentric distance. I don't see a reason to limit the capability to only concentric circles, as Abdullah mentioned circle to circle distance is the minimum distance between them.
Circle to point and circle to line are good additions too it makes the length.

**Comment by FlachyJoe** [comment]
So…
I get it working with a little trick for concentric case but I don't know how to deal with this situation:
image

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [799838307](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/799838307) was triggered at [4db3273](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/4db327377b3c72e90d1d5a01417f19786b6c26c3). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by PaddleStroke** [comment]
Great!

If the distance between centers is less than sum of radius (ie if one
circle is inside the other one) then maybe the sign should be reversed
because we probably don't want negative distances.


If so for one given distance, there would be 4 possible solutions,
1. both circles separated
2. circle 1 center is in circle 2
3. circle 2 center is in circle 1
4. both centers are in the other circle.

So the solver needs to apply constraint depending on these situations by
using different formulas.

Not sure if this makes sense as I haven't looked in details yet, but maybe
that will help you.

**Comment by abdullahtahiriyo** [comment]
Yes... the general case is more complex...sorry for that.

In addition, the circles can also be secant (one cutting the other).

I do not think it is possible to decide which distance is the one without actually comparing the four combinations (where the line joining the circles and its extension cuts both circles). So the four intersecting points.

There are at least three places where this needs to be calculated:
i) in CommandConstraints when calculating and applying the Acting Distance.
ii) in EditModeConstraintManager when drawing the labels
iii) at the solver when calculating the gradient and error function

for (i) and (ii), you may create a function in Mod/Sketcher/Gui/Utils.h to calculate the distance and return the points on the circle where the distance is minimum. These can be calculated using Vector3d.

Then, you need to calculate the same inside iii but using DeriVector.

One possibility for making this calculation is:
1) Calculating the inter-center vector using the centers V= C2-C1
2) Normalising the vector v = norm(V).
3) Then the four intersection points are:
i11 = C1+r1*v
i12 = C1-r1*v
i21 = C2+r2*v
i22 = C2-r2*v
4) Then the four possible lengths are:
L_i11_i21 = (i21-i11).Length()
L_i11_i22 = (i22-i11).Length()
L_i12_i21 = (i21-i12).Length()
L_i12_i22 = (i22-i21).Length()
5) Take the minimum of them
6) The intersection points leading to the minimum are the two points you are looking for, and the minimum length the distance.
7) For the solver constraint you have the derivector that lead to the minimum length, the error is basically the length of this vector minus the constraintvalue.

For those calculations using derivector, the ellipse constraint can serve as inspiration. For example: ```ConstraintInternalAlignmentPoint2Ellipse::errorgrad``` (as there emaj is a normalised vector that is afterwards scaled by the radius a, da).

Hope that helps.

**Comment by FlachyJoe** [comment]
@abdullahtahiriyo When circles intersect minimal distance is null I think this has to fall to _inner_ case.
I'll make this _centerI in circleJ_ test and treat this case specifically with no search for general mathematical solution.

**Comment by abdullahtahiriyo** [comment]
@FlachyJoe 

Yes, indeed, that is what I meant, minimal distance in the common perpendicular, i.e. in the line through both centers.

**Comment by abdullahtahiriyo** [comment]
Perhaps I did not manage to explain it well:
image

This is what I refer as the intersection points. There are 6 points, two are the centers, the rest are the intersection points. There are two per circle.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [800309241](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/800309241) was triggered at [f3ee5b7](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/f3ee5b76426954f8e578882e6c24525af68f4b70). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [800362861](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/800362861) was triggered at [9a344fb](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/9a344fb9ccfa4fb2e94f778eba9242a93d618482). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by PaddleStroke** [comment]
Have you looked into ViewProviderSketch ? There is a mode called 'STATUS_SKETCH_DragConstraint' that is activated in ViewProviderSketch::mouseButtonPressed.
Maybe it will point you in the correct direction.

**Comment by FlachyJoe** [comment]
@PaddleStroke Yes !
Found here:
https://github.com/FreeCAD/FreeCAD/blob/2bdef3a9a35ca75dd0d6664e2557c1d307e22031/src/Mod/Sketcher/Gui/ViewProviderSketch.cpp#L1389

I'd liked label in concentric case acts as radius but I can't get it working so it's always horizontal.

image

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [801849428](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/801849428) was triggered at [2bdef3a](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/2bdef3a9a35ca75dd0d6664e2557c1d307e22031). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [803250376](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/803250376) was triggered at [aa1f9ef](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/aa1f9ef3cb2202ec0cdac4e55c09b360c60383ab). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by FlachyJoe** [comment]
Rebased and squashed, ready for review.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [803251846](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/803251846) was triggered at [ba91e08](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/ba91e0885dfd093396b699f18811e30c7ad2f33d). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Sketcher/App/planegcs/Constraints.cpp [hunk 1]
```diff
@@ -2585,5 +2585,93 @@ double ConstraintEqualLineLength::grad(double *param)
...
+        if (err)
+            *err = *c2.rad - *c1.rad - d;
+        if (grad) {
+            if (param == c1.rad)
+                *grad = -1.0;
+            else if (param == c2.rad)
+                *grad = 1.0;
+            else
+                *grad = 0.0;
+        }
```
Comment (abdullahtahiriyo) [inline src/Mod/Sketcher/App/planegcs/Constraints.cpp:2655]: There are several issues with these differentials. as the gradients do not depend on the radii the circle does not resize. This you can best see by turning the constraint to reference (non-driving). You will see that the circle appears to keep following the constraint.

I will try to help you out here...

**Comment by abdullahtahiriyo** [comment]
Ok. These are some of the issues:

[Constraint_reference_constraint_not_working.webm](https://user-images.githubusercontent.com/5249162/224505271-f07a5c3b-5b4c-499e-9d27-c34a94595cb2.webm)
[Constraint_reference_would_not_pass_through_center.webm](https://user-images.githubusercontent.com/5249162/224505273-217623cf-32eb-46cf-a145-b00eb8b191d0.webm)
[Constraint_would_not_pass_through_center.webm](https://user-images.githubusercontent.com/5249162/224505274-41a7e63c-3ed6-4681-bc25-99dcaa206109.webm)
[Inner_circle_cannot_contain_center.webm](https://user-images.githubusercontent.com/5249162/224505276-3932da48-e5fe-4eff-ba23-5f7dac6cfcb7.webm)

**Comment by abdullahtahiriyo** [comment]
Another small detail I just realised. You should change the author of the commit. I understand you took some work from David. You may credit him in the commit message (with a link to his repo). But you have changed quite a lot of code and added new one.

**Comment by abdullahtahiriyo** [comment]
I am going to comment on the specific changes.

At this moment I do not have more time available. There is visualisation problem, that the label does not show the minimum distance (which corresponds to the value in the label):

image

so there, the 4 mm are on the other side. I think you can fix that one.

Then, we need to do quite some testing on this one. A constraint is a very important addition and we need to ensure we are getting it right.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [803563926](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/803563926) was triggered at [140c386](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/140c3868407405b215a05092cb895ee05ee900e5). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by abdullahtahiriyo** [comment]
Please have a look, fix the label issue and ensure it all works properly (to a fair amount of testing).

Then we will ask volunteers for a thorough testing.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [804970959](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/804970959) was triggered at [dbd4918](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/dbd4918cfee0dfaa4cbfec7c213073f106870df9). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745">pipeline status</a> for feature branch [PR_8745](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8745). Pipeline [804987298](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/804987298) was triggered at [24a09a9](https://github.com/FreeCAD/FreeCAD/pull/8745/commits/24a09a98b95411beb99a37d7edb0fdb021899bcc). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by FlachyJoe** [comment]
Replaced by #8896

**Pull request closed without merging**